src/lib.rs

```diff
@@ -6,34 +6,42 @@
 //!
 //! ## Consensus
 //!
-//! Consensus algorithms are fundamental to resilient, distributed systems such as decentralized
-//! databases and blockchains. Byzantine fault tolerant systems can reach consensus with a number
-//! of faulty nodes _f_ (including complete takeover by an attacker), as long as the total number
-//! _N_ of nodes is greater than _3 f_.
-//!
-//! The Honey Badger consensus algorithm is both Byzantine fault tolerant and asynchronous. It does
-//! not make timing assumptions about message delivery. An adversary can control network scheduling
-//! and delay messages without impacting consensus, and progress can be made in adverse networking
-//! conditions.
+//! A consensus algorithm is a protocol that helps a number of nodes agree on some data value.
+//! Byzantine fault tolerant systems can tolerate a number of faulty nodes _f_ (broken, or even
+//! controlled by an attacker), as long as the total number _N_ of nodes is greater than _3 f_.
+//! Asynchronous protocols do not make assumptions about timing: Even if an adversary controls
+//! network scheduling and can delay message delivery, consensus will still be reached as long as
+//! all messages are _eventually_ delivered.
 //!
+//! The Honey Badger consensus algorithm is both Byzantine fault tolerant and asynchronous. It is
+//! also modular, and the subalgorithms it is composed of are exposed in this crate as well, and
+//! usable separately.
 //!
-//! ## Crate Implementation
+//! Consensus algorithms are fundamental to resilient, distributed systems such as decentralized
+//! databases and blockchains.
 //!
-//! This protocol does not function in a standalone context, it must be instantiated in an
-//! application that handles networking.
 //!
-//! * The network must contain a number of nodes that are known to each other by some unique
-//! identifiers (IDs) and are able to exchange authenticated (cryptographically signed) messages.
+//! ## Usage
 //!
-//! * The user must define a type of _input_ - the _transactions_ - to the system and nodes must
-//! handle system networking.
+//! This crate is meant to be used as a component in a distributed application where the consensus
+//! problem arises. The application will usually run on different nodes, connected to each other
+//! via a network. The nodes give an input to the algorithm, exchange several messages, and
+//! eventually the algorithm returns an output, which is guaranteed to be the same in each correct
+//! node.
 //!
-//! * Messages received from other nodes must be passed into the instance, and messages produced by
-//! the instance sent to corresponding nodes.
+//! However, the `hbbft` crate only implements the abstract protocols, not the networking. It is
+//! the application's responsibility to serialize, sign and send the messages to the other nodes.
+//! That is why in addition to methods for giving input, the algorithms also have a
```

Comment:
'methods for giving input' is a little confusing, are you referring to the above 'nodes give an input' statement. Are you talking about methods for providing input to the nodes, or receiving info from the nodes - or both?  Is this referencing the`handle_input` method?

Reply:
Yes, I'm very unsure about this part. The problem is that a lot of it depends on what the user actually wants to do with the hbbft library, and what we mean by "node"—only the algorithm instance or the whole running program. (Probably the latter, in general.) I'm also trying to talk about all our algorithms and not a specific ones, which adds to the vagueness. :grimacing: 
By "giving input" I mean that the higher-level application logic has somehow obtained a value that it now passes to the algorithm instance.
For example:
* The application is Parity, and it has picked some of the pending transactions. It passes these as "input" to its instance of the `HoneyBadger` algorithm.
* A distributed chat application uses `Broadcast` to make sure all messages reach everyone. The user just typed a line, and it passes that text as input to the `Broadcast` algorithm.

But I wouldn't give an example at that point in the docs yet; they should start with a general formulation, I think.